services/sharedchannel/attachment.go

```diff
@@ -0,0 +1,188 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package sharedchannel
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/mattermost/mattermost-server/v5/mlog"
+	"github.com/mattermost/mattermost-server/v5/model"
+	"github.com/mattermost/mattermost-server/v5/services/remotecluster"
+)
+
+// postToAttachments returns the file attachments for a post that need to be synchronized.
+func (scs *Service) postToAttachments(post *model.Post, rc *model.RemoteCluster, lastSyncAt int64) ([]*model.FileInfo, error) {
+	infos := make([]*model.FileInfo, 0)
+
+	fis, err := scs.server.GetStore().FileInfo().GetForPost(post.Id, false, true, true)
+	if err != nil {
+		return nil, fmt.Errorf("could not get file info for attachment: %w", err)
+	}
+
+	for _, fi := range fis {
+		if scs.shouldSyncAttachment(fi, rc, lastSyncAt) {
+			infos = append(infos, fi)
+		}
+	}
+	return infos, nil
+}
+
+// postsToAttachments returns the file attachments for a slice of posts that need to be synchronized.
+func (scs *Service) shouldSyncAttachment(fi *model.FileInfo, rc *model.RemoteCluster, lastSyncAt int64) bool {
+	sca, err := scs.server.GetStore().SharedChannel().GetAttachment(fi.Id, rc.RemoteId)
+	if err != nil {
+		if _, ok := err.(errNotFound); !ok {
+			scs.server.GetLogger().Log(mlog.LvlSharedChannelServiceError, "error fetching shared channel attachment",
+				mlog.String("file_id", fi.Id),
+				mlog.String("remote_id", rc.RemoteId),
+				mlog.Err(err),
+			)
+		}
```

Comment:
I don't *really* want to, but if I don't then the attachment will get missed.  All the attachments have to go before the post (unless I refactor the app layer to handle attachments coming later) and if we skip this one on error I'd have to skip the post that owns it and retry them - get complicated.  If no record comes back due to error then we will redundantly upload the file again and it will error clean on the receiving side.  Just a bit of bandwidth wasted.